canonical/loki-k8s-operator
# Avoid setting data in relation departed event handler

## Issue
This PR is an trying to extract a minimal subset of work recently done by @rbarry82 @sed-i @balbirthomas and @Abuelodelanada in trying to make the integration tests for the Loki operator more stable. There were multiple issues as detailed below.

1. The integration tests use the cos-config operator. The cos-config operator loaded alert rules from the main github branch of this Loki charm. In order to test behaviour of faulty alert rules an incorrect alert rule was introduced into the set of test rules in this Loki charm. The cos-config operator recursively loaded all alert rules, in all the integration tests (including the fault alert rule). This was one cause of integration test failures. It was not possible to fix this failure without removing the cos-config operator because it was not possible to test the fix locally on a developer machine or on a PR branch. This was because the cos-config operator loaded the alert rules from the Loki operators main branch on github. 
2. Integration tests were also failing due to exceptions raised when the LokiPushApi provider tried to write to relation data when handling a relation departed event. This commit minimally refactors the Loki Operator to avoid this behaviour. Further refactor which may be useful have been documented in various github issues and are hence not implemented as part of this PR.
3. After replacing cos-config with loki-tester it was observed that charmcraft cached the state of the loki-tester. For example if an integration test inserted a faulty alert rule into loki-tester rebuilt and redeployed it, the deployed loki-tester was still missing the newly inserted alert rule. The same problem was observed if an alert rule was removed, loki-tester rebuilt and redeployed.
4. For quite some time (months) it was observed that errors used to be raised by python-libjuju and pytest-operator when the integration tests were tearing down models. However these errors did not lead to failing tests. However recent changes to these tools surfaced these failures as failures of the integration tests even though these errors occurred after the tests passed.
5. The loki operator sets its status to active even before its loki server workload is actually active and ready to service HTTP requests. As a result of this a prior integration test was using tenacity to query the loki server's status repeatedly till it passed. 
6. Juju sporadically results in a hung state when trying to install charms.
7. Juju sporadically results in a hung state when trying to remove applications and/or models.

## Solution
1. Replaced cos-proxy with a Loki Tester charm in all affected integration tests.
2. Minimally refactored Loki Operator to avoid writing application data in relation departed event. This refactor starts the process of moving away from using a single event handler of LokiPushApiProvider to handle multiple relation events with the consumer. Further refactors may be effected in subsequent commits such as those mentioned in issues https://github.com/canonical/loki-k8s-operator/issues/140, https://github.com/canonical/loki-k8s-operator/issues/146, https://github.com/canonical/loki-k8s-operator/issues/133, https://github.com/canonical/loki-k8s-operator/issues/149 and others

Specifically this change ensures the following

- There are three separate event handler for
   + upgrade event
   + relation departed event
   + relation changed event

- The upgrade event handler iterates over all relations with
   consumers and does the following
   + updates relation data for consumers
   + checks and updates all alert rules obtained from consumers

- On relation departed the LokiPushApi provider only updates
   its own alert rules. Prior to this commit the provider
   also set relation data for the consumer and checked alert
   rules. Neither of this are necessary. For example no new
   alert rule is obtained in a relation departed event hence
   there is no need to recheck the alert rules.

- This change also puts simple pebble can connect guards for
   two existing code blocks. Future refactors may make these
   guards more intelligent.

3. The integration test helper methods used to build loki-tester have been updated to ensure that `charmcraft clean` is invoked after inserting or removing additional alert rule files.
4. Both python-libjuju and pytest-operator were reverted to older versions known not to cause false positive integration test failures.
5. The use of tenacity was extended to another integration test that was checking if loki server is alive before it actually had the time to spin up. However an github issue ticket was created to address this issue properly in a future PR : https://github.com/canonical/loki-k8s-operator/issues/149. 
6. It is not possible to address the sporadic "hang on installing agent" through changes to a charm's integration testing code. This problem can only be resolved in Juju. It may be possible to detect this and restart the test but this is perhaps best left to human intervention to do so.
7. Exceptions raised by python-libjuju and pytest-operator on application removal are trapped in the tests where they are frequently observed. This allows the test to continue execution and check the test assertion to see if the tests still pass or not.

## Context
As described above


## Testing Instructions
All integration tests need to be run repeatedly to check they consistently pass.


## Release Notes
- Improved stability of integration tests
- Partially refactored Loki charm to have separate handler for each event.
